### ISSUE TITLE
Using numpy.random instead of the build-in random module

### ISSUE THREAD

**Issue by Kubasinska** [projectmesa/mesa#958]
**What's the problem this feature will solve?**
In my projects I have often implemented [numpy.Random](https://numpy.org/doc/stable/reference/random/generator.html) instead of using the random generator already implemented in the base class model.Model. The numpy.Random module offers much more functionality, an example: extracting a number from a poisson distribution. 

**Describe the solution you'd like**
What do you think? I can work on the implementation.

**Comment by jackiekazil** [comment]
This is an interesting question. If memory serves me right, numpy.random.seed() isn't thread safe. I don't believe we are using multi-threads at this time, but I could see a world where we would.... so I am on the fence on this. 

But you bring up some good points. 

This feels a little over the top, but I am wondering if we would want to support both and maybe make it a settings function that triggers which one is used. Thoughts? 

(I would also want one of the other maintainers to jump in.)

**Comment by Kubasinska** [comment]
I think a "switch" is the best option in case we want to implement this feature. As I have already said, I can implement it without any problem, therefore I am waiting for further updates.

**Comment by Corvince** [comment]
I am not really convinced by this. Sure, I think it often makes sense to use numpy.random, but I don't see the general benefit of adding such complexity and dependency (we only depend on numpy for the continous grid atm).

Isn't this readily solved by a single one-liner in ones `model.__init__` function a lá?

```python
self.random = np.random.default_rng(seed)
```

I think it would probably make sense to put this in the useful snippets section of the docs, but not more. (just my opinion)

**Comment by tcdejong** [comment]
I'm not concerned about thread safety. Given Python's GIL, concurrency primarily speeds up programs if it lets you avoid waiting for IO operations and the likes, if sharing state is required. That's not usually the bottleneck for ABMs. In other words, In other words, parallelizing the execution of _one_ model is hard. 

A much more realistic approach to concurrency is executing multiple configurations in parallel, which each execution being fully single threaded. This eliminates the difficulties of sharing state with multiprocessing, the GIL with multithreading, and provides more flexibility with proposals like this one. 

TLDR: I don't think thread-safety is a valid argument against this proposal.

**Comment by jackiekazil** [comment]
@tcdejong Agreed. But adding more complexity that is not clear can be more difficult for anyone trying to test out parallelizing the execution. 

I am currently leaning with @Corvince on the complexity argument, but could be talked out of it. 

@dmasad @tpike3 @rht -- any of you have strong feels one way or another?

**Comment by rht** [comment]
NetworkX has implemented both, with additionally including an integer as a seed [1] [2]. The seed is used a lot in many of the graph generator functions, and so they have made a decorator for it, called `py3_random_state`, which can be used by users too.

Extensive discussion at https://github.com/networkx/networkx/issues/1550 (note that this also points to several related issues).

I don't have a strong opinion yet of whether to keep the codebase simple or to support a variety of rng. Still reading the discussion!

[1] https://networkx.org/documentation/stable/reference/randomness.html
[2] https://github.com/networkx/networkx/blob/518798ba7aaa4bd980ec3822e13d0daf98892934/networkx/utils/misc.py#L421-L438

**Comment by rht** [comment]
Turns out that NetworkX's implementation is based on scikit-learn: https://github.com/scikit-learn/scikit-learn/blob/6b4f82433dc2f219dbff7fe8fa42c10b72379be6/sklearn/utils/validation.py#L869. I read that both NetworkX and sciki-learn only check for `insinstance` of `numpy.random.RandomState` (no builtin `random` in sight).

I think neither of the 2 libraries are using their rng within a multithreaded environment.

**Comment by tpike3** [comment]
My inclination is with @Corvince. Simpler is better and if users need to add the extra functionality in this case one like can provide it aided by a useful snippet.

**Comment by jackiekazil** [comment]
We talked about this during the dev meeting and after a discussion at this time, we don't want to switch out, because to do the override it is simple. We think there is value in the override and believe a good place to document this and share with others is our **Useful Snippets** section of our docs -- https://mesa.readthedocs.io/en/stable/useful-snippets/snippets.html#. We invite you to submit a PR. If you need any help doing this, please let us know.
